### PR TITLE
feat: [rfqm] Get transaction hash and store before submitting

### PR DESCRIPTION
**Pull request by rhinodavid**
~~**Requires 0xProject/protocol#310 to push before tests will pass**~~

**Closes MKR-96**

- Adds Ethers as a direct dependency of 0x API
- Adds function to sign and hash a transaction using ethers
- Adds function to submit a signed transaction still using web3wrapper
- Adds a Presubmit status to the RFQM Transaction Status enum

Now, when a worker is going to submit a transaction, the flow is:
1. Sign and hash the transaction offline
1. Save the transaction to the rfqm_transaction_submission table with a status of Presubmit
1. Submit the transaction to the mempool
1. Ensure the hash computed offline matches the hash from submission (a unit test is added to verify this works as expected, but the implementation also checks and logs an error if they differ)
1. Update the row in rfqm_transaction_submission with a status of Submitted

## Testing Instructions
Unit test added to ensure that the offline hash is the same as the hash from the rpc provider.

Checklist
 Update documentation as needed. Website Documentation PR:

 Prefix PR title with [WIP] if necessary.

 Add tests to cover changes as needed.

 Test changes on the staging environment with Matcha API Staging.

 SRA/Limit orders
 Swap endpoints
 Meta transaction endpoints
 Depth charts
For more information see 0x API Matcha smoke test runbook in Quip.

### CODE REVIEW

File: src/runners/http_rfqm_service_runner.ts
Comment (phil-ociraptor) [inline src/runners/http_rfqm_service_runner.ts:111]: there's a few functions in `rfq_blockchain_utils` such as `_getPathByIndex` that could reduce duplication of the `m/44'/60'/0'/0/` stuff. That suggests that we could create a new function called `createWallet` on `rfq_blockchain_utils`

File: src/services/rfqm_service.ts
Comment (phil-ociraptor) [inline src/services/rfqm_service.ts:1149]: to make this log easier to search upon, can we use the following syntax:
```
logger.error( {
 transactionHashFromSubmit,
 transactionHash: hash
}, "Mismatch between transaction hash calculated before submit and after submit");
```

This will allow us to search on the `transactionHash` and trace it's journey through the log messages

File: src/services/rfqm_service.ts
Comment (phil-ociraptor) [inline src/services/rfqm_service.ts]: nit: can we rename `transactionHashFromSubmit` as just `transactionHash`, just to make the log tracing easier

File: src/utils/rfq_blockchain_utils.ts
Comment (phil-ociraptor) [inline src/utils/rfq_blockchain_utils.ts:299]: nit: can we return `transactionHash` instead of the term `hash` - will keep it consistent whenever we use the term `transactionHash`

**Comment by phil-ociraptor** [comment]
First pass done - looks really good from a high level - just some nits

File: src/services/rfqm_service.ts
Comment (rhinodavid) [reply src/services/rfqm_service.ts:1149]: Oh good call

File: src/utils/rfq_blockchain_utils.ts
Comment (rhinodavid) [reply src/utils/rfq_blockchain_utils.ts]: sgtm I was going back and forth about this

File: src/services/rfqm_service.ts
Comment (rhinodavid) [reply src/services/rfqm_service.ts]: good call i was going back and forth about this one

File: src/runners/http_rfqm_service_runner.ts
Comment (rhinodavid) [reply src/runners/http_rfqm_service_runner.ts:111]: Ack. My preference is actually to rip out all those functions once we get rid of web3-wrapper. This line is concise and clear (I think) and we can get rid of a lot of code we have to maintain.

File: src/runners/http_rfqm_service_runner.ts
Comment (rhinodavid) [reply src/runners/http_rfqm_service_runner.ts]: Yep ethers specifically separates the functionality (which is nice imho). Unless you want to nix web3wrapper we have to have both

File: src/utils/rfq_blockchain_utils.ts
Comment (rhinodavid) [reply src/utils/rfq_blockchain_utils.ts]: The order format changes -- `gasPrice` gets removed and `max_priority_fee_per_gas, max_fee_per_gas` get added. I think we could hold off a little bit more on this but migrating is something we should look at. (My ledger wallet still doesn't seem to support these orders)

File: src/runners/http_rfqm_service_runner.ts
Comment (rhinodavid) [reply src/runners/http_rfqm_service_runner.ts:106]: if you really want to.. it's not reused so my preference is to keep it inline

File: src/utils/rfq_blockchain_utils.ts
Comment (rhinodavid) [reply src/utils/rfq_blockchain_utils.ts:224]: It gets caught and logged by `processRfqmJobAsync` just as it has before (tested this a BUNCH in ropsten lol)

File: src/services/rfqm_service.ts
Comment (rhinodavid) [reply src/services/rfqm_service.ts:1131]: `to` automatically gets set to the exchange proxy

File: src/utils/rfq_blockchain_utils.ts
Comment (rhinodavid) [reply src/utils/rfq_blockchain_utils.ts:265]: for the worker we do but added this for flexibility

File: src/utils/rfq_blockchain_utils.ts
Comment (phil-ociraptor) [inline src/utils/rfq_blockchain_utils.ts:277]: optional nit: can we add a comment calling out the renaming from `gas` to `gasLimit`?

File: test/rfq_blockchain_utils_test.ts
Comment (phil-ociraptor) [inline test/rfq_blockchain_utils_test.ts:281]: Can we add a test for `transformTxDataToTransactionRequestAsync`? Some logic in there that could do well having a test around:
- The fallback to address being the exchangeProxy
- The logic regarding mapping some of the fields
- The ethers fallback (not necessary to test this imo, but nice for completion - also see comment about possibly moving this logic out of this method)

File: src/utils/rfq_blockchain_utils.ts
Comment (phil-ociraptor) [inline src/utils/rfq_blockchain_utils.ts:49]: is this optional only because we're in the transition phase from web3wrapper to ethers? If so, can we drop the above comment and replace it with a TODO to make this non-optional once the migration is complete? Or just some message to remind us to remove the `undefined` union type at some point.

File: src/utils/rfq_blockchain_utils.ts
Comment (phil-ociraptor) [inline src/utils/rfq_blockchain_utils.ts:271]: this logic was a little unexpected (was expecting a pure fn). Thoughts on replicating this functionality by offering a `getWalletAddress` fn on `RfqBlockchainUtils`? Then a caller can stitch together this logic as needed

File: src/runners/http_rfqm_service_runner.ts
Comment (phil-ociraptor) [reply src/runners/http_rfqm_service_runner.ts:106]: I prefer to move it into a constant, but will leave it up to you. (i.e. definitely not a blocker)

**Review by phil-ociraptor — APPROVED** [review]
Looks really good! Added some more minor comments but this looks good from my side, especially after tested on Ropsten

File: src/utils/rfq_blockchain_utils.ts
Comment (rhinodavid) [reply src/utils/rfq_blockchain_utils.ts:49]: Hmmmm.. it's only necessary if you're using `RfqBlockchainUtils` as a Worker (the API version doesn't need to sign/submit transactions). We could make it mandatory, but then we'd have to mock it for tests where it's not used. What do you think?

File: src/utils/rfq_blockchain_utils.ts
Comment (rhinodavid) [reply src/utils/rfq_blockchain_utils.ts:271]: Yeah Daniel didn't like it either.. maybe we just say it has to be provided

File: src/utils/rfq_blockchain_utils.ts
Comment (rhinodavid) [reply src/utils/rfq_blockchain_utils.ts:277]: done

File: src/utils/rfq_blockchain_utils.ts
Comment (rhinodavid) [reply src/utils/rfq_blockchain_utils.ts]: ok done

File: test/rfq_blockchain_utils_test.ts
Comment (rhinodavid) [reply test/rfq_blockchain_utils_test.ts:281]: good call done

File: src/utils/rfq_blockchain_utils.ts
Comment (rhinodavid) [reply src/utils/rfq_blockchain_utils.ts]: removed

**Pull request merged**
